Performance: Response

bertjwregeer opened the pull request:
Based on comments made in #285 I took a look at seeing what could be done to get some of the speed back in the synthetic test that `howareyou` provides.

In that ticket there is a patchset that brings it back to within 2k rps in the benchmark. However the maintenance burden goes up drastically, and I don't think the additional burden is worth the cost.

The new and updated `Response.__init__` is more correct, and has more code re-use by calling various properties/functions rather than trying to do it all. There is a sacrifice, but ultimately WebOb and it's `Response` object are not going to be the bottleneck, especially because outside of the synthetic test that attempts to return a `Response` object as fast as possible user code is going to dominate in cases where WebOb is used.

----

Update: After talking about the way that `Response.__init__` works with @mmerickel , a bunch of mis-features that used to exist have been removed:

If the user provides a headerlist, Response.__init__ will no longer attempt to set the default Content-Type, nor will __init__ add a charset to any existing Content-Type.

The Content-Type will only be set if no headerlist is provided, and the HTTP status code allows for a body. The charset will be set only if there is no charset on the Content-Type provided.

    resp = Response(headerlist=[], content_type='text/plain', charset='UTF-8')

will have a headerlist that is: `resp.headerlist == [('Content-Length', '0')]`

    resp = Response(status='204 No Content')

will have a headerlist that is: `resp.headerlist == []`

`__init__` will now also ignore body/app_iter if the status code has no body (100-199, 204, 304)

    resp = Response(status='204 No Content', app_iter=[b'test'])

Will not have a `Content-Length`, and `resp.body` will return `b''`. If you would like to violate HTTP standards, you may of course do:

    resp = Response(status='204 No Content')
    resp.body = b'test'

`__init__` won't let you create an invalid response, what you do with the returned response however is up to you.

Closes #285

bertjwregeer commented:
```
# master before this branch
Benchmark : hello
Frameworks: pyramid
                ms     rps  tcalls  funcs
pyramid       9430   10604     144     73

# Fixes applied
(howareyou)alexandra:howareyou xistence$ howareyou -b hello -f pyramid
Benchmark : hello
Frameworks: pyramid
                ms     rps  tcalls  funcs
pyramid       8112   12327     123     72
```

bertjwregeer commented:
```
# WebOb 1.6.1
Benchmark : hello
Frameworks: pyramid
                ms     rps  tcalls  funcs
pyramid       5299   18870      74     58

# WebOb with this branch
Benchmark : hello
Frameworks: pyramid
                ms     rps  tcalls  funcs
pyramid       5051   19797      71     56
```

bertjwregeer commented:
@mmerickel Could you take a hard look at this and make sure I didn't miss anything. I added a bunch of new tests because while everything was "covered" there was no tests for `default_body_encoding` and `default_charset`

bertjwregeer commented:
@faassen this one is for you.

bertjwregeer commented:
```
Benchmark : hello
Frameworks: pyramid
                ms     rps  tcalls  funcs
pyramid       4829   20706      69     54
```

Cleaned up `_abs_headerlist`. 69 tcalls...

```
   Ordered by: internal time
   List reduced from 65 to 10 due to restriction <10>

   ncalls  tottime  percall  cumtime  percall filename:lineno(function)
   100000    0.838    0.000    4.670    0.000 router.py:65(handle_request)
   100000    0.589    0.000    0.704    0.000 response.py:181(__init__)
   100000    0.427    0.000    5.776    0.000 router.py:166(invoke_subrequest)
   100000    0.318    0.000    7.016    0.000 run.py:42(<lambda>)
   100000    0.297    0.000    0.418    0.000 response.py:1241(__call__)
   100000    0.295    0.000    0.567    0.000 traversal.py:632(__call__)
   300000    0.264    0.000    0.313    0.000 interface.py:518(__hash__)
   100000    0.250    0.000    6.600    0.000 router.py:214(__call__)
   100000    0.214    0.000    0.945    0.000 urldispatch.py:80(__call__)
   200000    0.197    0.000    0.323    0.000 decorator.py:34(__get__)
```

WebOb is listed twice. `Response.__init__` and `Response.__call__`.

bertjwregeer commented:
It may be possible to cut down the `__call__` more, AFAIK the following status code are allowed to have a `Location` header on them:

- 201
- 301
- 302
- 303
- 305
- 307

Sending a `Location` header on any others is not allowed, and from bug reports across the internet it causes issues with many clients that would be on the receiving end of that.

bertjwregeer commented on webob/response.py:
```diff
@@ -225,43 +222,66 @@ def __init__(self, body=None, status=None, headerlist=None, app_iter=None,
+        encoding = None
+        code_has_body = (
+            self._status[0] != '1' and
+            self._status[:3] not in ('204', '304')
```
Add `205` to this small list. It is like a `204` but it requires the client to reset the document view (i.e. clear a form).

mmerickel commented on webob/response.py:
```diff
@@ -225,43 +222,66 @@ def __init__(self, body=None, status=None, headerlist=None, app_iter=None,
-                encoding = self.charset
+                # Fall back to trying self.charset if encoding is not set. In
+                # most cases encoding will be set to the default value.
+                encoding = encoding or self.charset
```
Shouldn't `default_body_encoding` be used here before raising an error?

bertjwregeer replied:
I don't think so, I would love to hear arguments for/against it... but `default_body_encoding` is strictly for `.text` to me. 

`__init__` won't let you create a `204 No Content` with a body for example (because setting Content-Length on such a response is not allowed by the HTTP spec), `__init__` will do it's best to give you a valid `Response`. If you make changes/set invalid options on the `Response` after the fact, then so be it.

There is no way to influence `default_body_encoding` in `__init__` so you end up using the default all of the time. I'd recommend encoding before it reaches `Response`. Changing this would also be b/w incompatible, if you set the `content_type` to `image/jpg` in older versions of WebOb you couldn't provide a text body because there would be no charset set.

Reading through the 1.6.1 code though, it seems if you set `charset` and `content_type` it would happily create `image/jpg; charset=utf-8` and then use the charset if passed in. So I guess I could use the provided `charset` as if it were the `default_body_encoding`, even-though it is not added the `content_type`.

If you do `Response(text='sometext')` then `__init__` will call the `.text` attribute and `default_body_encoding` will be used to set the `body` up though...

mmerickel replied:
afaict passing `body` as a `text_type` is similar to invoking `.text` (and in my view setting `.body` with text data should be equivalent to setting `.text` with the data). It looks to me that right now `Response(body=u'foo')` works but `r = Response(); r.body = u'foo'` does not.

bertjwregeer replied:
`body` has accepted only `bytes` (or it's equivalent `str` in Py2 for a VERY long time): https://github.com/Pylons/webob/blob/0.8/webob/__init__.py#L1445

It would be an API asymmetry to change that, because `r.body` will always return bytes, which may be unexpected. I feed it text, I get back bytes from the same property?

That's what `.text` is for.

The fact that `Response(body=u'foo')` works is that in 0.9 `__init__` was updated to call `unicode_body`: https://github.com/Pylons/webob/blob/0.9/webob/__init__.py#L1364 when the "body" provided was unicode. I assume it was a convenience thing, because `.body` was not updated to match. `.unicode_body` sometime in its history got renamed to `.text`, and instead of calling `self.unicode_body` vs `self.body` the features from it were inlined into `__init__` probably for speed reasons, which brings us back to where we are today.

Unfortunately the name "body" is being re-used for something that accepts both `bytes`/`str` and `unicode`.

----

That being said, falling back to `default_body_encoding` seems like it could be a sane thing to do, and matches what was done in the past.

Side note: Ian at one point did only set the default `charset` on text types:

https://github.com/Pylons/webob/blob/0.9/webob/__init__.py#L1357

Somewhere between that version and 1.6.x it was lost, and thus if you didn't explicitly provide a `charset` providing `u'foo'` as a `body` parameter would have failed due to not having a valid charset.

bertjwregeer replied:
WebOb 1.6.1 behaviour:

```
>>> Response(content_type='image/jpg', body=u'test',charset=None).body
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/Users/xistence/Projects/webob/webob/response.py", line 139, in __init__
    "You cannot set the body to a text value without a "
TypeError: You cannot set the body to a text value without a charset
>>> Response(content_type='image/jpg', body=u'test').body
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/Users/xistence/Projects/webob/webob/response.py", line 139, in __init__
    "You cannot set the body to a text value without a "
TypeError: You cannot set the body to a text value without a charset
>>> Response(content_type='image/jpg', body=u'test',charset='UTF-16').body
b'\xff\xfet\x00e\x00s\x00t\x00'
```
If we used `default_body_encoding` both of the first two examples would succeed and no error would be raised.

bertjwregeer replied:
If we want to break those first two examples, I can make that change, so that it looks more like `.text` and falls back to `default_body_encoding`. I am -0 on it personally.

mmerickel replied:
Okay, thanks for the clarification. I think that if `body` is unicode then the goal should be to raise. The inconsistency bothers me. There's a lot of work being put into getting things correct in the `__init__` but my experience is usually affecting the `Response` **after** init via things like `request.response.foo = bar` or just creating a response object before configuring it so if there's things that I'm losing by not doing it all at `__init__` time, I'd like to know.

bertjwregeer replied:
Changing it to always raise of the `body` is unicode would break a lot of existing applications that depend on the behaviour.

mmerickel replied:
Not interested in deprecating it? Usually you have seem happy to aim for consistency so I'm a little surprised.

bertjwregeer replied:
There is nothing that you are losing by not doing it at `__init__` time, after `__init__` time there are however less seatbelts. For example:

```
res = Response(status='204 No Content')
res.body = b'A body'
```

Would create a response that had:

```
assert res.body == b'A body'
assert res.status == '204 No Content'
assert res.headers['Content-Length'] == 6
```

Which is very much not allowed by the HTTP spec. a `204 No Content` response is not allowed to have any message-body. This also means no `Content-Type`/`Content-Length` is allowed. `Content-Length: 0` is also invalid, and various HTTP libraries will throw the entire response away as invalid.

bertjwregeer replied:
As for deprecating it... I'm all for it, but that means `Response('body')` no longer works. Are we ready to deal with that fall-out?

```
>>> Response('test')
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/Users/xistence/Projects/webob/webob/response.py", line 308, in __init__
    raise TypeError("Expected bytes value but received text.")
TypeError: Expected bytes value but received text.
```

I'm generally all for it, except this time I have a feeling it's going to break MANY applications.

mmerickel replied:
Well in your example you didn't specify `body=` so it feels like you could set `.text` instead of `.body` since it was left ambiguous. For the case of `Response(body='body')` I can see the concern.

bertjwregeer replied:
The first parameter that `__init__` takes is body. If I made that `.text` instead, things would go badly too:

(Assumes that the first param to `__init__` is `.text`)

```
>>> Response(b'test')
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/Users/xistence/Projects/webob/webob/response.py", line 335, in __init__
    setattr(self, name, value)
  File "/Users/xistence/Projects/webob/webob/response.py", line 647, in _text__set
    "(not %s)" % type(value))
TypeError: You can only set Response.text to a unicode string (not <class 'bytes'>)
```

mmerickel replied:
```
def __init__(self, *args, **kw):
    if len(args) > 0:
        if instance(args[0], bytes):
            kw['body'], args = args[0], args[1:]
        elif isinstance(args[1], text_type):
            kw['text'], args = args[0], args[1:]
    if len(args) > 0:
        raise ValueError('unknown args')
```

mmerickel commented on webob/response.py:
```diff
@@ -225,43 +222,66 @@ def __init__(self, body=None, status=None, headerlist=None, app_iter=None,
+                    if (
+                        content_type is self.default_content_type or
+                        _content_type_has_charset(content_type)
+                    ):
```
Does the `default_content_type` imply that it needs a charset?

bertjwregeer replied:
It does in the case of the default_content_type we ship with, but not if it was overridden. I think I'd want to change this to `content_type is 'text/html'` since that is our default, and it saves us a function call and a bunch of checking in the default case.

mmerickel replied:
What about baking the charset into the `default_content_type` instead of handling them separately in this case? Then `default_content_type` is unambiguous, and `default_charset` is only used if someone provides a partial content-type.

bertjwregeer replied:
It wouldn't allow someone to sub-class `Response` and override only the `default_charset` or set it to `None` for instance.

mmerickel replied:
Sorry I don't follow. It seems to me if you trust `default_content_type` as fully specified then it simplifies things instead of asking them to work together.

bertjwregeer replied:
```
    class DefaultCharset(Response):
        default_charset = 'UTF-16'

    res = DefaultCharset()
    assert res.content_type == 'text/html'
    assert res.charset == 'UTF-16'
    assert res.headers['Content-Type'] == 'text/html; charset=UTF-16'
```

Would no longer change the charset of the `content_type` if `default_content_type` was set to `text/html; charset=UTF-8`. To change the `default_charset` in that case, you'd have to also override the `default_content_type`.

The test would have to become:

```
    class DefaultCharset(Response):
        default_content_type = 'text/html; charset=UTF-16'
        default_charset = 'UTF-16'

    res = DefaultCharset()
    assert res.content_type == 'text/html'
    assert res.charset == 'UTF-16'
    assert res.headers['Content-Type'] == 'text/html; charset=UTF-16'
```

mmerickel replied:
In your first example, what is `default_content_type`?

bertjwregeer replied:
It is the `default_content_type` from `Response`: `text/html`.

bertjwregeer replied:
```
    class DefaultCharset(Response):
        default_charset = 'UTF-16'

    res = DefaultCharset()
    assert res.content_type == 'text/html'
    assert res.charset == 'UTF-16'
    assert res.headers['Content-Type'] == 'text/html; charset=UTF-16'

    assert DefaultCharset.default_content_type == 'text/html'
    assert DefaultCharset.default_charset == 'UTF-16'
```

with the change you'd recommended:

```
    assert DefaultCharset.default_content_type == 'text/html; charset=UTF-8'
    assert DefaultCharset.default_charset == 'UTF-16'
```

mmerickel replied:
Yes, if for some reason you wanted the `default_charset` to not be UTF-8 then that'd be what you do.

bertjwregeer replied:
Making that change doesn't have much effect on performance at all. I'd rather leave it as is.

mmerickel replied:
This is a usability change. I wasn't arguing for it in favor of performance.

mmerickel commented on webob/response.py:
```diff
@@ -1248,7 +1270,15 @@ def conditional_response_app(self, environ, start_response):
-        headerlist = self._abs_headerlist(environ)
+
+        # Build headerlist, if there is a Location header, make it absolute
+        headerlist = [
```
note something about how this `_abs_headerlist` code is copied around / inlined from somewhere for future editors... possibly even leaving the original impl in place with tests, and a note about where it is inlined to... and the inline points should point back to the canonical impl

bertjwregeer replied:
Good idea, but:

I've been thinking about removing this entirely. Making the `Location` absolute really doesn't provide much especially since all clients since HTTP 1.0 support relative Location's. It's doing extra work that feels unnecessary.

Thoughts?

mmerickel replied:
I'm not against it but it will be a library regression for a bunch of people - at least breaking a bunch of test suites until it's fixed. The behavior itself **should** work, but may expose some malformed non-browser clients or anyone handling location headers themselves.

bertjwregeer replied:
Brought back `_abs_headerlist`, and added an optimization to the default path to only call it when the HTTP Status Code allows for a `Location` header. If someone has a `Location` header on any other status codes they are not within spec, and it won't be made absolute. If this breaks test suites, then they need to get fixed.

bertjwregeer commented on tests/test_response.py:
```diff
@@ -1241,3 +1243,48 @@ def test_cache_expires_set_zero_then_nonzero():
+    res = DefaultCharset()
+    assert res.content_type == 'text/html'
+    assert res.charset == 'UTF-16'
+    assert res.headers['Content-Type'] == 'text/html; charset=UTF-16'
```
I am not sure I follow...

mmerickel replied:
I was commenting that this test confirms that if you set charset **after** content_type, it still appends the charset as expected.

bertjwregeer replied:
Ah, that's not what this test checks.

This test verifies that sub-classing `Response` with a new `default_charset` actually uses the `default_charset` provided by the sub-class and not the original.

mmerickel replied:
Oh I guess I read that wrong. Does what I specified work?

bertjwregeer replied:
```
r = Response()
r.content_type = 'application/foo'
r.charset = 'UTF-16'

assert r.headers['Content-Type'] == 'application/foo; charset=UTF-16'
```

However:

```
r = Response()

# Change charset to UTF-16
r.charset = 'UTF-16'

assert r.headers['Content-Type'] == 'text/html; charset=UTF-16'

# Change the content_type, this removes charset and raises a warning
r.content_type = 'application/foo'

assert r.headers['Content-Type'] == 'application/foo'
```

bertjwregeer replied:
Yes. This change was made because of this:

```
res = Response()
res.content_type = 'image/jpg'
```

used to keep the content-type parameters, which also meant stuff like:

```
assert res.headers['Content-Type'] == 'image/jpg; charset=UTF-8'
```

This was changed in WebOb 1.6.0. Currently other content-type parameters are copied from the old to the new:

```
res = Response(content_type='application/foo; level=1; charset=UTF-8')
res.content_type = 'image/jpeg'

assert res.headers['Content-Type'] == 'image/jpeg; level=1'
```

Which is going to go away in the future. Content-Type parameters shouldn't be kept when moving from one Content-Type to another.

mmerickel commented on tests/test_response.py:
```diff
@@ -1288,3 +1288,26 @@ def test_set_content_type():
+    assert res.content_type == 'image/jpeg'
+    assert res.charset is None
+
+def test_conten_type_has_charset():
```
`s/conten/content`

mmerickel commented on tests/test_response.py:
```diff
@@ -1288,3 +1288,26 @@ def test_set_content_type():
+    res = Response(content_type='application/foo; charset=UTF-8', body=text_(b'test'))
+    assert res.body == b'test'
+    assert res.content_type == 'application/foo'
+    assert res.charset == 'UTF-8'
```
Can I expect `assert res.headerlist = [('Content-Type', 'application/foo; charset=UTF-8')]` here? I was slightly surprised to see `res.content_type` not include the charset.

mmerickel replied:
Wait didn't you just say that I can't set `.body` with unicode and get back bytes? So if I set `response.content_type = 'application/foo; charset=UTF-8'` then `assert response.content_type == 'application/foo'` will pass?

bertjwregeer replied:
I've updated these tests to make it more clear.

bertjwregeer replied:
Yes. `r.content_type` can be set with a full `content_type` string and everything works correctly. As in `r.headers['Content-Type'] == 'application/foo; charset=UTF-8'`. `r.content_type` only returns the content-type not the content-type with parameters.

Certain properties have some really bad assymetric API's. Unfortunately they have been that way for a long time, can't change that now.

mmerickel commented on webob/response.py:
```diff
@@ -1287,13 +1296,11 @@ def __call__(self, environ, start_response):
-            in self._headerlist
-        ]
+        # If the HTTP Status code might have a Location header, try to make it absolute
+        if self._status[:3] in ('201', '301', '302', '303', '305', '307'):
```
For starters this doesn't support 308. In general, I don't like this change. It's actually fairly difficult to figure out what responses support Location... you're only checking right now for the responses that **require** Location. It's complicated.

bertjwregeer replied:
Setting Location on a `200 OK` HTTP Status code is not allowed for example, and will cause some intermediaries to change the status code to a 302 (Apache's mod_proxy IIRC), and others will send a 500 Server Error back to the remote client (Squid). Various HTTP checkers (for example OpenStack has some neat ones) also complain loudly. You can't just set Location on stuff willy nilly. Same thing with Content-Length for example, you can't set that on responses without a body.

I can add 308 to the list, or I can drop the optimization completely. It doesn't make that big of a difference.

bertjwregeer replied:
```
(howareyou)alexandra:howareyou xistence$ howareyou -b hello -f pyramid
Benchmark : hello
Frameworks: pyramid
                ms     rps  tcalls  funcs
pyramid       4723   21173      70     55
```

Dropped the extra optimization, not worth it.

bertjwregeer commented:
@mmerickel I'm sorry, but I don't think this is the right place to discuss the API issues in `Response`.

At this point I am happy with the changes in this PR, are there any other major concerns that you have? Outside issues with the API as it stands?

mmerickel commented:
Well where do you want me to go from here? You made API changes in this PR that is labeled performance. I need to discuss them somewhere. Do you want to make a new performance PR that is **only** a performance change?

bertjwregeer commented:
Wait... what API change did I make? The only change made was that charset is no longer added to a Content-Type that is provided in the `headerlist`.

No other API changes have been made.

You are discussing and asking about `.body` and `.text`, as well as `.content_type` but those have not changed behaviour at all in this PR.

bertjwregeer commented:
```
____________________________________________________________________ test_204_has_no_body ____________________________________________________________________

    def test_204_has_no_body():
        res = Response(status='204 No Content')
        assert res.body == b''
>       assert res.content_length is None
E       assert 0 is None
E        +  where 0 = <Response at 0x102ae3390 204 No Content>.content_length

tests/test_response.py:1244: AssertionError
___________________________________________________________________ test_204_app_iter_set ____________________________________________________________________

    def test_204_app_iter_set():
        res = Response(status='204', app_iter=[b'test'])
>       assert res.body == b''
E       assert b'test' == b''
E         Left contains more items, first extra item: 116
E         Use -v to get the full diff

tests/test_response.py:1249: AssertionError
_________________________________________________ test_doesnt_raise_with_charset_content_type_has_no_charset _________________________________________________

    def test_doesnt_raise_with_charset_content_type_has_no_charset():
        res = Response(content_type='image/jpeg', body=text_(b'test'), charset='utf-8')
        assert res.body == b'test'
        assert res.content_type == 'image/jpeg'
>       assert res.charset is None
E       assert 'utf-8' is None
E        +  where 'utf-8' = <Response at 0x104b768d0 200 OK>.charset

tests/test_response.py:1278: AssertionError
=========================================================== 3 failed, 1095 passed in 5.92 seconds ============================================================
```

Test failures on `master` that are working on `perf/responses`.

mmerickel commented:
Well the Location header is no longer normalized for certain response codes. Honestly the rest of it is just super hard to keep track of because the original behavior isn't exactly documented with regards to `content_type`, `charset`, `default_content_type` and `default_charset`. As you showed several of the tests on master with regards to those arguments have broken, so something changed and I've been trying to figure out what the intended behavior is. Feel free to merge this if you want to discuss the API issues elsewhere, but I don't know the original behavior well enough to evaluate this with regards to just the performance optimizations (and through no fault of anyone, certainly not you, I don't have a lot of faith in simply the test suite to show what has actually changed).

bertjwregeer commented:
> Well the Location header is no longer normalized for certain response codes.

Nope, I removed that code. It is always normalized, for all HTTP response codes.

> Honestly the rest of it is just super hard to keep track of because the original behavior isn't exactly documented with regards to content_type, charset, default_content_type and default_charset.

The only tests that fail are the "no body" ones. Which I added (on this new branch perf/responses). `204 No Content` shouldn't have a `Content-Length` and shouldn't have a `body`.

The only charset one that fails is when the user passes in a `headerlist` with a `Content-Type` that doesn't have a `charset` already on it. We discussed this on IRC, you said if someone passes a `headerlist` then there should be no changing what is in that `headerlist` since it is a low level api.

I agreed to that.

Currently on `master` if you pass in a `headerlist` with a `Content-Type` it will add the `charset` to it automatically.

WebOb 1.6.1/webob master:

```
res = Response(headerlist=[('Content-Type', 'text/html')], charset='UTF-8')

assert res.headers['Content-Type'] == 'text/html; charset=UTF-8'
```

WebOb perf/responses:

```
res = Response(headerlist[('Content-Type', 'text/html')], charset='UTF-8')

assert res.headers['Content-Type'] == 'text/html'
```

once again, we discussed this on IRC. `headerlist` is a lower level API.

In short, only NEW tests for new behaviour I've added to make sure that behaviour can't regress in the future are failing on master. Tests from master pass without issues on perf/responses.

Those same tests that fail on `master` that are on `perf/response` also fail when testing on 1.6-branch. So there is no changes there.

If you want me to pull the "no body" changes, I can do that, and make a new branch for the couple of line changes.

> Feel free to merge this if you want to discuss the API issues elsewhere,

Yes, I don't think that this is the place.

> but I don't know the original behavior well enough to evaluate this with regards to just the performance optimizations (and through no fault of anyone, certainly not you, I don't have a lot of faith in simply the test suite to show what has actually changed).

The test suite isn't half bad at showing regressions. Most of the changes made in master for the 1.7 series so far have required changes in the test suite, and I spent a lot of time going through the tests when I converted it to py.test.

mmerickel commented:
Sounds good, thanks for being so patient with me. Let's discuss the content-type and charset stuff elsewhere.

bertjwregeer commented:
No problem. I am all for making things better. If you want to put together a bunch of questions I'll do my best to answer them and any history surrounding them.
